Create a `TypeRegistry` and enable injection of `Type` instances

GromNaN opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no (deprecation)
| Fixed issues | Closes #2818

#### Summary

- Introduce a new class `TypeRegistry` that replace the static method of the `Type` class. The name is the same as [DBAL `TypeRegistry`](https://github.com/doctrine/dbal/blob/4.4.x/src/Types/TypeRegistry.php), with the same methods (except `override` because `register` creates or replace the type, "upsert" is a MongoDB specificity 😉)
- Deprecate all the static methods of the `Type` class
- Inject the `TypeRegistry` into the `Configuration` and add the public method `getTypeRegistry()`. This method is used by `UnitOfWork`, `DocumentPersister` and the aggregation builder.
- Inject the `TypeRegistry` into `ClassMetadata` instances.

The Bundle PR: https://github.com/doctrine/DoctrineMongoDBBundle/pull/963

GromNaN committed:
Replace Type static functions with an instance of TypeRegistry

Enable having a distinct list of types for each instance of DocumentManager in 3.0

Enable injecting type instances constructed with parameters,
which is necessary for generic embedded object transformations.

GromNaN committed:
Trigger deprecations in legacy Type static methods

GromNaN committed:
Inject the TypeRegistry into ClassMetadata

GromNaN committed:
Enable injection of Type instance into the TypeRegistry

GromNaN commented on src/DocumentManager.php:
```diff
@@ -205,6 +210,21 @@ public function resolveClassName(string $className): string
+        $registry           ??= $this->typeRegistry;
+
+        if ($registry !== $this->typeRegistry) {
+            throw new LogicException('Type registry is already set and cannot be modified. A single TypeRegistry instance is allowed, this will change in MongoDB ODM 3.0.');
```
There is a design flow with Symfony `cache:clear` because the container is created twice in the same process. So we have 2 distinct `TypeRegisty` that are created.

This static variable is set to maintain BC for the static methods of the `Type` class.

GromNaN committed:
Use a shared instance of TypeRegistry by default when none is injected

GromNaN committed:
Update docs

alcaeus commented on docs/en/reference/custom-mapping-types.rst:
```diff
@@ -79,24 +79,28 @@ Restrictions to keep in mind:
-    // Registers a type without checking whether it was already registered
-    Type::registerType('date_immutable', \My\Project\Types\DateTimeWithTimezoneType::class);
+    // Initialize DocumentManager with the TypeRegistry
+    $dm = DocumentManager::create($client, $config, null, $types);
```
Nit: I'd suggest a different variable name (`$typeRegistry`)

```suggestion
    $typeRegistry = new TypeRegistry();

   // Adds a type. This results in an exception if type with given name is already registered
    $typeRegistry->register('date_with_timezone', \My\Project\Types\DateTimeWithTimezoneType::class);

   // Overrides a type. This results in an exception if type with given name is not registered
    $typeRegistry->register('date_immutable', \My\Project\Types\DateTimeWithTimezoneType::class);

    // Initialize DocumentManager with the TypeRegistry
    $dm = DocumentManager::create($client, $config, typeRegistry: $typeRegistry);
```

alcaeus commented on docs/en/reference/custom-mapping-types.rst:
```diff
@@ -182,7 +186,8 @@ Register the type in your bootstrap code::
 
-    Type::addType(Money::class, App\MongoDB\Types\MoneyType::class);
+    $types = new TypeRegistry();
+    $types->register(Money::class, App\MongoDB\Types\MoneyType::class);
```
```suggestion
    $typeRegistry = new TypeRegistry();
    $typeRegistry->register(Money::class, App\MongoDB\Types\MoneyType::class);
```

GromNaN replied:
Even better if I inject an instance?
```suggestion
    $types = new TypeRegistry();
    $types->register(Money::class, new App\MongoDB\Types\MoneyType());
```

alcaeus replied:
Absolutely!

alcaeus commented on src/DocumentManager.php:
```diff
@@ -144,7 +145,7 @@ class DocumentManager implements ObjectManager
      * and uses the given Configuration.
      */
-    protected function __construct(?Client $client = null, ?Configuration $config = null, ?EventManager $eventManager = null)
+    protected function __construct(?Client $client = null, ?Configuration $config = null, ?EventManager $eventManager = null, private ?TypeRegistry $typeRegistry = null)
```
Is there a reason why you didn't add this to the configuration class similar to metadata factories etc. I'm not sure what I prefer, so I'm totally leaving this as is and starting to extract such services from the configuration class.

GromNaN replied:
Indeed, this is also consistent with the configuration class. It's all very mixed up. The configuration class can also have methods for adding a type... which would allow `TypeRegistry` to be made internal.

GromNaN replied:
Also, I don't like adding a method to the `DocumentManager`. So the `Configuration` class makes sense.

alcaeus commented on src/DocumentManager.php:
```diff
@@ -955,4 +957,9 @@ public function getClassNameForAssociation(array $mapping, $data): string
         return $mapping['targetDocument'];
     }
+
+    public function getTypes(): TypeRegistry
```
```suggestion
    public function getTypeRegistry(): TypeRegistry
```

alcaeus commented on src/Types/TypeRegistry.php:
```diff
@@ -0,0 +1,169 @@
+        if ($type instanceof Type) {
+            $this->typesMap[$name]    = $type::class;
+            $this->typeObjects[$name] = $type;
+        } else {
```
I suppose this case handles somebody registering a type by its FQCN. We should still ensure that the type implements the `Type` interface.

GromNaN replied:
I've added a check in the `register` method to ensure the class doesn't have a constructor.

alcaeus commented on src/Types/TypeRegistry.php:
```diff
@@ -0,0 +1,169 @@
+            throw InvalidTypeException::invalidTypeName($name);
+        }
+
+        return $this->typeObjects[$name] ??= (new ReflectionClass($this->typesMap[$name]))->newInstanceWithoutConstructor();
```
We make the assumption that a type has no dependencies. This is fine for the time being, but we should consider changing this pattern and removing the option to provide just a FQCN for a type, instead requiring people to add types using a type instance.

alcaeus commented on src/Types/TypeRegistry.php:
```diff
@@ -0,0 +1,169 @@
+     *
+     * @internal
+     */
+    public function fromVariable(mixed $variable): ?Type
```
I would rename this to better indicate that we're guessing a type, and that we're expecting a native value and not a database value.

```suggestion
    public function guessTypeFromValue(mixed $variable): ?Type
```

In future, we could consider adding a separate interface that allows types to indicate whether they can handle a certain value.

GromNaN replied:
Method renamed with your good suggestion.

> In future, we could consider adding a separate interface that allows types to indicate whether they can handle a certain value.

That's delicate as we need to have the same detection from property type and from values.

alcaeus reviewed (commented):
Initial review done. I like the idea of a type registry, but it once again shows that we really need to clean up the dependency mess that we currently have.
The only concern I have is exposing the `convertToDatabase` method from DocumentPersister. I think we should mark it as `@internal` to ensure people don't treat it as part of the public API (unless the whole persister is already marked as such).

GromNaN committed:
Update the documentation

GromNaN committed:
Forbids the injection of a type per class name if it has a constructor

GromNaN commented on src/Persisters/DocumentPersister.php:
```diff
@@ -1087,7 +1086,7 @@ public function prepareQueryOrNewObj(array $query, bool $isNewObj = false): arra
      * @return mixed
      */
-    private function convertToDatabaseValue(string $fieldName, $value, ?ClassMetadata $class = null)
+    public function convertToDatabaseValue(string $fieldName, $value, ?ClassMetadata $class = null)
```
> The only concern I have is exposing the `convertToDatabase` method from `DocumentPersister`. I think we should mark it as `@internal` to ensure people don't treat it as part of the public API (unless the whole persister is already marked as such).

@alcaeus the class `DocumentPersister` is already all `@internal`.

GromNaN committed:
Rename TypeRegistry::guessTypeFromValue

GromNaN committed:
Validate the registered class in TypeRegistry

greg0ire commented on src/Types/TypeRegistry.php:
```diff
@@ -0,0 +1,179 @@
+/**
+ * The TypeRegistry is responsible for managing the mapping types supported.
+ */
+class TypeRegistry
```
Should it be final?

GromNaN replied:
Then I would have to create an interface since it's not internal. Which is more strict for BC than a regular class.

greg0ire replied:
> Then I would have to create an interface since it's not internal.

Do you? Is every type supposed to be replacable/mockable? I think it's OK to marks something as `final` and postpone the decision of making it extensible vs having an interface to when a need for extensibility actually emerges. For something that looks like an map with extra methods, I'm not convinced extensibility is needed.

Feel free to ignore this if you have a hard rule about this in this project.

alcaeus replied:
Agree with making it final. We don't need to mock it for internal tests (yet), and people should definitely be mocking it for their own tests considering it's a registry they can create and inject as they see fit. We can always make the class non-final in a future minor release if necessary.

GromNaN replied:
I've already a use-case for extending the TypeRegistry: in the bundle I add lazy-loading of type instances from the container: https://github.com/doctrine/DoctrineMongoDBBundle/pull/963/changes#diff-b090a1e775d0625ed9280a9140328e118d6ad52fafb71464bf7246ac937e8442

GromNaN replied:
I'll rework this TypeRegistry class to also support `ContainerInterface`, and add a `TypeProvider` interface. As done in https://github.com/doctrine/dbal/pull/7342

greg0ire commented:
Some commits look like they should be squashed together.

GromNaN commented:
> Some commits look like they should be squashed together.

Yes, my plan was to have 2 separate commits to ease with rebase and merges, but the implementation changed a lot and that will not help anymore. I'll squash-merge the PR as we usually do on the MongoDB-ODM project.

GromNaN committed:
Uptimize Date compaison in UOW::computeOrRecomputeChangeSet

MongoDB has 1 ms of precision, we use UTCDateTime object to normalize the timezone and skip the microseconds

GromNaN committed:
Move TypeRegistry to the Configuration class

GromNaN committed:
Create TypeRegistry when the DM is not used

GromNaN committed:
Revert date comparison to use field type from ClassMetadata

GromNaN committed:
Make TypeRegistry::getMap non-internal, so that it can be overridden in the bundle

alcaeus reviewed (commented):
Code LGTM. Do you want to add the deprecations to the UPGRADE document now or do we do that in the 3.0.x branch when we remove deprecated functionality?

GromNaN committed:
Add upgrade instructions

GromNaN committed:
Fix CS

GromNaN committed:
Fix shard key field name bug and move upgrade instructions to 2.17

- Fix DocumentPersister::prepareShardKeyQuery() passing $mapping['name']
  (DB field name) to getFieldType() which expects the PHP property name;
  use $mapping['fieldName'] instead, consistent with all other call sites
- Move TypeRegistry upgrade instructions from UPGRADE-2.16.md to the new
  UPGRADE-2.17.md, and fix method names in the migration table
  (get/has/register/getMap instead of the old static method names)

GromNaN committed:
Add test for shard key with custom DB field name

Covers the bug where getFieldType() received the DB field name
($mapping['name']) instead of the PHP property name ($mapping['fieldName']),
causing type conversion to be skipped when the two names differ.

GromNaN committed:
Document that ClassMetadata must be loaded through ClassMetadataFactory

Direct deserialization of ClassMetadata has never been supported, as the
factory is responsible for restoring runtime dependencies after cache hits.

GromNaN committed:
Fix TypeRegistry::register() to allow types with inherited no-arg constructors

Replace method_exists() check (which incorrectly blocked types inheriting a
no-arg constructor from a parent) with a try/catch on instantiation. This
correctly rejects only classes whose constructor has required parameters,
and as a bonus caches the created instance so get() never re-instantiates.

GromNaN committed:
Add test for Configuration::setTypeRegistry() immutability

GromNaN committed:
Fix version in deprecation messages

GromNaN commented:
Rebased on 2.17.x and reworked to follow doctrine/dbal#7342, so both projects expose the same concepts under the same names. The description above has been rewritten accordingly.

What changed since the previous state of this PR:

- Added the `TypeProvider` interface (`get()`, `has()`, `Traversable<string, Type>`). It is now what `Configuration` accepts and returns, so `TypeRegistry` is one implementation rather than the only option.
- `TypeRegistry` is `final` and implements `TypeProvider` and `IteratorAggregate`. Built-in types moved to a class constant, and the constructor takes a map of instances or class names layered on top.
- Types can be lazy-loaded from a PSR-11 container, given a map of type names to service IDs. The container is never queried during construction, nor by `has()`. This is what lets the bundle wire a service locator instead of building every custom type at boot.
- `getMap()` is replaced by `getIterator()`, so stopping the iteration early no longer instantiates the remaining types.
- `guessTypeFromValue()` and `convertToDatabaseValue()` moved out of the registry into a separate `@internal` `TypeGuesser`. These two are ODM specifics with no DBAL equivalent, and keeping them off the interface makes `TypeProvider` implementable by third parties.
- `register()` no longer instantiates a class just to validate it. The constructor is inspected by reflection instead, so registering by class name stays lazy.
- Tests inject a registry scoped to the test `Configuration`, which removes the reflection reset of the shared instance in `tearDown()`.

I kept the immutability guard on `Configuration::setTypeRegistry()`, which DBAL does not have. Swapping the registry after `ClassMetadataFactory` has injected it into already loaded metadata would leave those instances pointing at the old one. I intend to propose the same guard for DBAL rather than drop it here.

`psr/container` is a `require-dev` plus a `suggest`, not a hard requirement, since the union type and the `catch` clause are only resolved when a container is actually passed.

This also fixes three tests that were already failing on the branch: the constructor rule relaxed in fc4d6e5 was never reflected in `CustomTypeTest`, and two deprecation messages still said 2.16.
